js/core/template_manager.js

```diff
@@ -0,0 +1,277 @@
+import $ from './renderer';
+import { isDefined, isFunction, isRenderer } from './utils/type';
+import { findBestMatches, noop } from './utils/common';
+import { extend } from './utils/extend';
+import { Error, log } from './errors';
+import { getElementOptions, normalizeTemplateElement } from './utils/dom';
+import devices from './devices';
+import { Template } from './templates/template';
+import { TemplateBase } from './templates/template_base';
+import { FunctionTemplate } from './templates/function_template';
+import { EmptyTemplate } from './templates/empty_template';
+import { ChildDefaultTemplate } from './templates/child_default_template';
+import { camelize } from './utils/inflector';
+
+const TEXT_NODE = 3;
+const ANONYMOUS_TEMPLATE_NAME = 'template';
+const TEMPLATE_SELECTOR = '[data-options*="dxTemplate"]';
+const TEMPLATE_WRAPPER_CLASS = 'dx-template-wrapper';
+const DEPRECATED_WIDGET_NAMES = ['button', 'tabs', 'dropDownMenu'];
+const DX_POLYMORPH_WIDGET_TEMPLATE = new FunctionTemplate(({ model, parent }) => {
+    let widgetName = model.widget;
+    if(!widgetName) return $();
+
+    const widgetElement = $('<div>');
+    const widgetOptions = model.options || {};
+
+    if(DEPRECATED_WIDGET_NAMES.some(deprecatedWidgetName => deprecatedWidgetName === widgetName)) {
```

Comment:
```suggestion
    if(DEPRECATED_WIDGET_NAMES.has(widgetName)) {
```
`Set` better fits here